product-tours/tour-pytorch/Tour-with-PyTorch.ipynb

```diff
@@ -0,0 +1,593 @@
+    "    optimizer.step()\n",
+    "\n",
+    "    # Log image predictions\n",
+    "    if batch_idx % 50 == 1:\n",
+    "        for image, prediction in zip(data, outputs):\n",
+    "            description = '\\n'.join(['class {}: {}'.format(i, pred)\n",
+    "                                     for i, pred in enumerate(F.softmax(prediction, dim=0))])\n",
+    "            neptune.log_image('predictions',\n",
+    "                              image.squeeze(),\n",
+    "                              description=description)\n",
+    "\n",
+    "    if batch_idx == PARAMS['iterations']:\n",
+    "        break"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": [
+     "header"
+    ]
+   },
+   "source": [
+    "### Log model weight to experiment"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": [
+     "comment"
+    ]
+   },
+   "source": [
+    "Trained model can be stored in Neptune and associated with the experiment. Use `log_artifact` method to do it."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": [
+     "code"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "torch.save(model.state_dict(), 'model_dict.pth')\n",
+    "neptune.log_artifact('model_dict.pth')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": [
+     "tests"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# tests\n",
+    "exp = neptune.get_experiment()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "AW4MM4qj5ErX",
+    "tags": [
+     "header"
+    ]
+   },
+   "source": [
+    "#### Stop Neptune experiment after training"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": [
+     "comment"
+    ]
+   },
+   "source": [
+    "Below method is necessary only for notebooks users. In the Python scipts experiment is closed automatically when script finishes."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": [
+     "tests"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# tests\n",
+    "# check logs\n",
+    "correct_logs_set = {'batch_loss', 'batch_acc', 'predictions'}\n",
+    "from_exp_logs = set(exp.get_logs().keys())\n",
+    "\n",
+    "assert correct_logs_set == from_exp_logs, '{} - incorrect logs'.format(exp)\n",
+    "\n",
+    "# check parameters\n",
+    "assert set(exp.get_parameters().keys()) == set(PARAMS.keys()), '{} parameters do not match'.format(exp)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 71
+    },
+    "colab_type": "code",
+    "id": "GvDSBSrOx-R4",
+    "outputId": "667d1b2f-cdef-44f0-f505-832205dea223",
+    "tags": [
+     "code"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "neptune.stop()"
```

Comment:
I think it is safer to run neptune.stop() before running the tests (.stop() will make all the logs from async threads synced). That said I don't think it makes a difference with this particular example and test.

Reply:
ok